ballista/rust/scheduler/src/state/task_manager.rs

```diff
@@ -37,16 +37,19 @@ use ballista_core::serde::{AsExecutionPlan, BallistaCodec};
 use datafusion::physical_plan::ExecutionPlan;
 use datafusion::prelude::SessionContext;
 use datafusion_proto::logical_plan::AsLogicalPlan;
-use log::{debug, info, warn};
+#[cfg(not(test))]
+use log::info;
+use log::{debug, error, warn};
 use rand::distributions::Alphanumeric;
 use rand::{thread_rng, Rng};
-use std::collections::{HashMap, HashSet};
+use std::collections::HashMap;
 use std::default::Default;
 use std::sync::Arc;
 use tokio::sync::RwLock;
 use tonic::transport::Channel;
 
 type ExecutorClients = Arc<RwLock<HashMap<String, ExecutorGrpcClient<Channel>>>>;
+type ExecutionGraphCache = Arc<RwLock<HashMap<String, Arc<RwLock<ExecutionGraph>>>>>;
 
 #[derive(Clone)]
 pub struct TaskManager<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> {
```

Comment:
This requires changes to the executors as well right? Currently the executor will just have a single connection to the scheduler but it will need to accept the scheduler URL as part of the task definition for this to work correctly.

Reply:
You are correct. For this part, I'll propose another PR.

Reply:
I think we would have to implement that as part of this PR or else any multi-scheduler deployment would break.

Reply:
I don't think we can achieve multiple scheduler now, since there's only one scheduler in the executor.

Reply:
The executor only has one scheduler connection but the scheduler's can (currently) assign tasks to any executor since the executor pool is a shared global state. So they would currently report status to only one scheduler instance (unless the scheduler endpoint was pointing to a load balancer or something) but there is no guarantee that they would have received the task assignment from that scheduler. 

I think it should be relatively straightforward though. The `TaskAssignment` would just need a scheduler URL in it and the executor would report status back to that URL (instead of maintaining a single global connection).

Reply:
If there's only one scheduler in the executor, how can the executors register themselves to multiple schedulers?

Reply:
And I don't think it's a good idea to include scheduler url to the task definition. We can achieve the scheduler url in the grpc request like the scheduler does for the executors.

Reply:
The executor only has one scheduler connection but when it registers it that one scheduler will add the executor to the global state. At that point, all schedulers will see that resource and be able to launch tasks on it. For pull-based scheduling this will be fine but for push-based scheduling it would break down.

Reply:
I need to get back to speed with the current architecture before I can help with this review. I won't have time to do that until this weekend, unfortunately. 

It would be good if we had a doc in this repo with the architecture we are working towards to make sure we are all on the same page. I can maybe try and start that if nobody beats me to it.

Reply:
For push-based task scheduling, if all of the executors report task update to only one scheduler, when the whole cluster is busy, I think there's only one scheduler will be busy scheduling tasks and other schedulers will be idle due to there're no `ExecutorReservation` left for them.

Reply:
I'm OK to add another commit of the multiple scheduler change in the executors to this PR. However, it may take a few days.

Reply:
What we do is configure the scheduler endpoint in the executor to be the kubernetes service endpoint. So this will load balance executor registrations across the schedulers  but any given scheduler should maintain a persistent connection to only the scheduler it registered with.

Reply:
Got it. I'll propose the commit in early next week.